File: stack/infix_to_postfix_linkedlist.py
# File Name: infix_to_postfix_ll.py

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Stack:
    def __init__(self):
        self.top = None

    def push(self, ch):
        new_node = Node(ch)
        new_node.next = self.top
        self.top = new_node

    def pop(self):
        if self.top is None:
            return None
        ch = self.top.data
        self.top = self.top.next
        return ch

    def peek(self):
        if self.top is None:
            return None
        return self.top.data

    def is_empty(self):
        return self.top is None

def precedence(ch):
    if ch in ['+', '-']:
        return 1
    if ch in ['*', '/']:
        return 2
    return 0

infix = input("Enter infix expression: ")
stack = Stack()
postfix = ""

for ch in infix:
    if ch.isalnum():
        postfix += ch
    elif ch == '(':
        stack.push(ch)
    elif ch == ')':
        while stack.peek() != '(':
            postfix += stack.pop()
        stack.pop()
    else:
        while (not stack.is_empty()) and precedence(stack.peek()) >= precedence(ch):
            postfix += stack.pop()
        stack.push(ch)

while not stack.is_empty():
    postfix += stack.pop()

print("Postfix Expression:", postfix)

File: stack/InfixToPostfixLinkedList.java
/* File Name: InfixToPostfixLL.java */

import java.util.Scanner;

class Node {
    char data;
    Node next;

    Node(char data) {
        this.data = data;
        this.next = null;
    }
}

class Stack {
    Node top = null;

    void push(char ch) {
        Node newNode = new Node(ch);
        newNode.next = top;
        top = newNode;
    }

    char pop() {
        if (top == null) return '\0';
        char ch = top.data;
        top = top.next;
        return ch;
    }

    char peek() {
        if (top == null) return '\0';
        return top.data;
    }

    boolean isEmpty() {
        return top == null;
    }
}

public class InfixToPostfixLL {

    static int precedence(char ch) {
        if (ch == '+' || ch == '-') return 1;
        if (ch == '*' || ch == '/') return 2;
        return 0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        Stack s = new Stack();
        String infix, postfix = "";

        System.out.print("Enter infix expression: ");
        infix = sc.nextLine();

        for (char ch : infix.toCharArray()) {
            if (Character.isLetterOrDigit(ch)) {
                postfix += ch;
            }
            else if (ch == '(') {
                s.push(ch);
            }
            else if (ch == ')') {
                while (s.peek() != '(')
                    postfix += s.pop();
                s.pop();
            }
            else {
                while (!s.isEmpty() && precedence(s.peek()) >= precedence(ch))
                    postfix += s.pop();
                s.push(ch);
            }
        }

        while (!s.isEmpty())
            postfix += s.pop();

        System.out.println("Postfix Expression: " + postfix);
        sc.close();
    }
}
